htmd/smallmol/chemlab/builder.py

```diff
@@ -0,0 +1,406 @@
+        """
+        Adds the hydrogens to fill up the valence of the atoms. With onlyExplicit is possible to add only the polar ones
+
+        Parameters
+        ----------
+        onlyExplicit: bool
+            If True, only the polar hydrogens will be added
+            Default: False
+        """
+
+        pT = self.periodictable
+        sm = self.smallmol
+
+        # polar hydrogens are the ones bonded to hetero atoms (no carbon)
+        if onlyExplicit:
+            heavy_sel = ' '.join(_hetero_atoms)
+        else:
+            heavy_sel = ' '.join(_heavy_atoms)
+
+        atomidxs = sm.get( 'idx', 'element {}'.format(heavy_sel))
+        formalcharges = sm.get('formalcharge', 'element {}'.format(heavy_sel))
+
+        missing_atoms = np.array([pT.getMissingValence(aId, sm, fch) for aId, fch in zip(atomidxs, formalcharges)])
+        ids = np.where(missing_atoms > 0)
+        heavyToFill = atomidxs[ids]
+        numHydrogens = missing_atoms[ids]
+        #print("missing hydrogens: ", numHydrogens)
+        h_atoms = []
+
+        for n_h, a in zip(numHydrogens, heavyToFill):
+            for i in range(n_h):
+                new_atom = pT.getAtom('H', a)
+                h_atoms.append(new_atom)
+
+        self._addAtoms(h_atoms)
+
+
+    def removeHydrogens(self, removePolars=True, removeNonPolars=True):
+        """
+        Removes the hydrogens. Is it possible to choose the type of hydrogens to remove: polar and not polar
+
+        Parameters
+        ----------
+        removePolars: bool
+            If True, the polar hydrogens will be removed
+            Default: True
+        removeNonPolars: bool
+            If True, the non polar hydrogens will be removed
+            Default: True
+        """
+
+        ## TODO
+        ## Fix the the kekulization problem whem removing hydrogen from NH in aromatic system.
+        sm = self.smallmol
+        hydrogensIdx =  sm.get('idx', 'element H')
+
+        neighborsIdx = np.concatenate(sm.neighbors[hydrogensIdx])
+        neighbors = sm.element[neighborsIdx]
+        hydrogensPolarIdx = hydrogensIdx[np.where( neighbors != 'C' )]
+        hydrogensNonPolarIdx = hydrogensIdx[np.where(neighbors == 'C')]
+
+
+        hydrogensIdx = np.array([],dtype=np.int)
+        if removePolars:
+            hydrogensIdx = np.append( hydrogensIdx, hydrogensPolarIdx)
+        if removeNonPolars:
+            hydrogensIdx = np.append(hydrogensIdx, hydrogensNonPolarIdx)
+
+        self._removeAtoms(hydrogensIdx)
+
+    def _addAtoms(self, atoms):
+        """
+        Adds the atoms one by one from the list of atoms data passed
+
+        Parameters
+        ----------
+        atoms: list
+            List of the atom data to add
+        """
+
+        sm = self.smallmol
+
+        #c = 1
+        for a in atoms:
+            n_atom = sm.numAtoms
+           # print('placing {} to {}'.format(n_atom, a['attachTo']))
+            sm.__dict__['idx'] = np.append(sm.idx, n_atom)
```

Comment:
There is no need to use `__dict__` here. Just do `sm.idx` etc.

Reply:
Here I guess the append is ok since you would need to convert to a list first and back again.

Reply:
done it. Should I have to do like that in all the parts where I used self.__dict__ ?

Reply:
`self.__dict__` should only be used when you don't know the name of the attribute. I.e. 
```python
x = 'names'
x = 'resid'
self.__dict__[x]
```
If you know you want the resid just do directly `self.resid`. 

Practically to be more correct, I was schooled recently by @Acellera/david that using the `__dict__` is not the best way. There are in python two functions called `getattr` and `setattr` which work as `setattr(sm, x, value)` and are probably the most correct way of doing it. I will try using that from now on but I don't want to make you change all that now. Too much work and it's my fault for doing this the wrong way to begin with

Reply:
I will keep it in mind and in the next future I'll refactor this by considering ```setattr``` and ```getattr```.